### ISSUE TITLE
Compilation errors on Ubuntu 20.04 / 18.04

### ISSUE THREAD

**Issue by iomartin** [poseidonos/poseidonos#31]
On a fresh Ubuntu 20.04.3, I got errors compiling both the library and Poseidon-OS.

Following the steps described on the README:

```bash
git clone https://github.com/poseidonos/poseidonos.git
cd poseidonos/script
sudo ./pkgdep.sh
cd ../lib
sudo ./build_lib.sh
cd ../script
sudo ./build_ibofos.sh
```

This is the error from building the library:

```
(...)
/home/martin/poseidonos/lib/grpc                                                                                                                                                                                                                                                                                              
/home/martin/poseidonos/lib/grpc/cmake/build /home/martin/poseidonos/lib/grpc                                                                                                                                                                                                                                                 
-- The C compiler identification is GNU 9.3.0                                                                                                                                                                                                                                                                                 
-- The CXX compiler identification is GNU 9.3.0                                                                                                                                                                                                                                                                               
-- Check for working C compiler: /usr/bin/cc                                                                                                                                                                                                                                                                                  
-- Check for working C compiler: /usr/bin/cc -- works                                                                                                                                                                                                                                                                         
-- Detecting C compiler ABI info                                                                                                                                                                                                                                                                                              
-- Detecting C compiler ABI info - done                                                                                                                                                                                                                                                                                       
-- Detecting C compile features                                                                                                                                                                                                                                                                                               
-- Detecting C compile features - done                                                                                                                                                                                                                                                                                        
-- Check for working CXX compiler: /usr/bin/c++                                                                                                                                                                                                                                                                               
-- Check for working CXX compiler: /usr/bin/c++ -- works                                                                                                                                                                                                                                                                      
-- Detecting CXX compiler ABI info                                                                                                                                                                                                                                                                                            
-- Detecting CXX compiler ABI info - done                                                                                                                                                                                                                                                                                     
-- Detecting CXX compile features                                                                                                                                                                                                                                                                                             
-- Detecting CXX compile features - done                                                                                                                                                                                                                                                                                      
-- Looking for pthread.h                                                                                                                                                                                                                                                                                                      
-- Looking for pthread.h - found                                                                                                                                                                                                                                                                                              
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD                                                                                                                                                                                                                                                                                    
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed                                                                                                            
-- Looking for pthread_create in pthreads                                                                                                                                                                                                                                                                                     
-- Looking for pthread_create in pthreads - not found                                                                                                                                                                                                                                                                         
-- Looking for pthread_create in pthread                                                                                                                                                                                                                                                                                      
-- Looking for pthread_create in pthread - found                               
-- Found Threads: TRUE                                                                                                                                         
-- Could NOT find c-ares (missing: c-ares_DIR)                                                                                                                 
-- Found c-ares: /usr/include (found suitable version "1.15.0", minimum required is "1.13.0")                                                                  
-- Found Protobuf: /usr/local/lib/libprotobuf.a;-lpthread (found version "3.15.8")                                                                             
-- Found RE2 via CMake.                                                                                                                                        
CMake Error at /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:146 (message):                                                                
  Could NOT find OpenSSL, try to set the path to OpenSSL root folder in the                                                                                    
  system variable OPENSSL_ROOT_DIR (missing: OPENSSL_CRYPTO_LIBRARY                                                                                            
  OPENSSL_INCLUDE_DIR)                                                                                                                                         
Call Stack (most recent call first):                                                                                                                           
  /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:393 (_FPHSA_FAILURE_MESSAGE)                                                               
  /usr/share/cmake-3.16/Modules/FindOpenSSL.cmake:447 (find_package_handle_standard_args)                                                                      
  cmake/ssl.cmake:66 (find_package)                                                                                                                            
  CMakeLists.txt:257 (include)                                                                                                                                 
                                                                                                                                                               
                                                                                                                                                               
-- Configuring incomplete, errors occurred!                                                                                                                    
See also "/home/martin/poseidonos/lib/grpc/cmake/build/CMakeFiles/CMakeOutput.log".                                                                            
See also "/home/martin/poseidonos/lib/grpc/cmake/build/CMakeFiles/CMakeError.log".                                                                             
make[3]: warning: -j4 forced in submake: resetting jobserver mode.                                                                                             
make[3]: *** No targets specified and no makefile found.  Stop.                                                                                                
make[3]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.                                                                              
make[3]: *** No rule to make target 'install'.  Stop.                                                                                                          
/home/martin/poseidonos/lib/grpc                                               
GRPC installation success                                                                                                                                      
[ 92%] Built target grpc                                                                                                                                       
-- spdk download command succeeded.  See also /home/martin/poseidonos/lib/spdk-prefix/src/spdk-stamp/spdk-download-*.log
```

I thought that this was due to the missing package `libssl-dev`, but I checked and it was already installed.

And this is the error from `sudo ./build_ibofos.sh`:

```
/home/martin/poseidonos/src/allocator/context_manager/rebuild_ctx/rebuild_ctx.cpp: In constructor ‘pos::RebuildCtx::RebuildCtx(pos::TelemetryPublisher*, pos::AllocatorAddressInfo*)’:
/home/martin/poseidonos/src/allocator/context_manager/rebuild_ctx/rebuild_ctx.cpp:69:31: error: ‘*<unknown>.pos::RebuildCtx::tp’ is used uninitialized in this function [-Werror=uninitialized]
   69 | : RebuildCtx(tp, nullptr, info)

```

**Comment by mjlee34** [comment]
Unfortunately, PoseidonOS can only be built on Ubuntu-18.04. We recommend you to try PoseidonOS on Ubuntu-18.04.
We should consider supports for the latest Ubuntu.

**Comment by iomartin** [comment]
The same error happened on a fresh Ubuntu 18.04 VM when running `build_lib.sh`, and two new errors happened on `build_ibofos.sh`


```
$ sudo ./build_lib.sh
(...)
~/poseidonos/lib/grpc
~/poseidonos/lib/grpc/cmake/build ~/poseidonos/lib/grpc
-- The C compiler identification is GNU 7.5.0
-- The CXX compiler identification is GNU 7.5.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features 
-- Detecting C compile features - done  
-- Check for working CXX compiler: /usr/bin/c++    
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info                     
-- Detecting CXX compiler ABI info - done              
-- Detecting CXX compile features 
-- Detecting CXX compile features - done 
-- Looking for pthread.h                      
-- Looking for pthread.h - found              
-- Looking for pthread_create                     
-- Looking for pthread_create - not found         
-- Looking for pthread_create in pthreads         
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE                      
-- Could NOT find c-ares (missing: c-ares_DIR)         
-- Found c-ares: /usr/include (found suitable version "1.14.0", minimum required is "1.13.0") 
-- Found Protobuf: /usr/local/lib/libprotobuf.a;-lpthread (found version "3.15.8") 
-- Found RE2 via CMake.          
CMake Error at /usr/share/cmake-3.10/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
  Could NOT find OpenSSL, try to set the path to OpenSSL root folder in the
  system variable OPENSSL_ROOT_DIR (missing: OPENSSL_CRYPTO_LIBRARY                                                                                            
  OPENSSL_INCLUDE_DIR)                     
Call Stack (most recent call first):                                                                                                                           
  /usr/share/cmake-3.10/Modules/FindPackageHandleStandardArgs.cmake:378 (_FPHSA_FAILURE_MESSAGE)
  /usr/share/cmake-3.10/Modules/FindOpenSSL.cmake:390 (find_package_handle_standard_args)                         
  cmake/ssl.cmake:66 (find_package)      
  CMakeLists.txt:257 (include)                                                                                                                                 
                                                                               
                                                                               
-- Configuring incomplete, errors occurred!         
See also "/home/martin/poseidonos/lib/grpc/cmake/build/CMakeFiles/CMakeOutput.log".
See also "/home/martin/poseidonos/lib/grpc/cmake/build/CMakeFiles/CMakeError.log".
make[3]: warning: -jN forced in submake: disabling jobserver mode.
make[3]: *** No targets specified and no makefile found.  Stop.
make[3]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
make[3]: *** No rule to make target 'install'.  Stop.
~/poseidonos/lib/grpc
```

```
$ sudo ./build_ibofos.hs
(...)
[ 48%] Performing update step for 'filebench'                                                                                                                  
CMake Error at /home/martin/poseidonos/lib/filebench-prefix/src/filebench-stamp/filebench-update-.cmake:16 (message):                                                                                                                                                                                                         
  Command failed: 1                                                                                                                                                                                                                                                                                                           
                                                                                                                                                                                                                                                                                                                              
   '/usr/bin/cmake' '-P' '/home/martin/poseidonos/lib/filebench-prefix/tmp/filebench-gitupdate.cmake'                                                                                                                                                                                                                         
                                                                                                                                                                                                                                                                                                                              
  See also                                                                                                                                                                                                                                                                                                                    
                                                                                                                                                                                                                                                                                                                              
    /home/martin/poseidonos/lib/filebench-prefix/src/filebench-stamp/filebench-update-*.log                                                                                                                                                                                                                                   
                                                                                                                                                                                                                                                                                                                              
                                                                                                                                                                                                                                                                                                                              
CMakeFiles/filebench.dir/build.make:94: recipe for target 'filebench-prefix/src/filebench-stamp/filebench-update' failed                                                                                                                                                                                                      
make[2]: *** [filebench-prefix/src/filebench-stamp/filebench-update] Error 1                                                                                                                                                                                                                                                  
CMakeFiles/Makefile2:493: recipe for target 'CMakeFiles/filebench.dir/all' failed                                                                                                                                                                                                                                             
make[1]: *** [CMakeFiles/filebench.dir/all] Error 2                                                                                                                                                                                                                                                                           
make[1]: *** Waiting for unfinished jobs....

(...)

make -C telemetry/telemetry_config clean
clean
make -C telemetry/telemetry_config/configs clean
clean
make -C dpdk_wrapper clean
clean
make -C memory_checker clean
clean
make: *** No rule to make target '/home/martin/poseidonos/lib/spdk/build/lib/libspdk_bdev_malloc.a', needed by '/home/martin/poseidonos/bin/poseidonos'.  Stop.
```

**Comment by iomartin** [comment]
Running `build_lib.sh` a second time now passes the ssl error, but fails in two new places

```
$ sudo ./build_lib.sh
(...)
[ 47%] Performing update step for 'filebench'                                                                                                                                                                                                                                                                                 
CMake Error at /home/martin/poseidonos/lib/filebench-prefix/src/filebench-stamp/filebench-update-.cmake:16 (message):                                          
  Command failed: 1                                                                                                                                            
                                                                                                                                                               
   '/usr/bin/cmake' '-P' '/home/martin/poseidonos/lib/filebench-prefix/tmp/filebench-gitupdate.cmake'                                                          
                                                                                                                                                               
  See also                                                                                                                                                     

    /home/martin/poseidonos/lib/filebench-prefix/src/filebench-stamp/filebench-update-*.log


CMakeFiles/filebench.dir/build.make:94: recipe for target 'filebench-prefix/src/filebench-stamp/filebench-update' failed
make[2]: *** [filebench-prefix/src/filebench-stamp/filebench-update] Error 1
CMakeFiles/Makefile2:493: recipe for target 'CMakeFiles/filebench.dir/all' failed
make[1]: *** [CMakeFiles/filebench.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....

(...)

-- Installing: /usr/local/lib/pkgconfig/gpr.pc
-- Installing: /usr/local/lib/pkgconfig/grpc.pc
-- Installing: /usr/local/lib/pkgconfig/grpc_unsecure.pc
-- Installing: /usr/local/lib/pkgconfig/grpc++.pc
-- Installing: /usr/local/lib/pkgconfig/grpc++_unsecure.pc
~/poseidonos/lib/grpc
GRPC installation success
[ 61%] Built target grpc
Makefile:83: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by mjlee34** [comment]
Can you print the result of the following?: $cat /home/martin/poseidonos/lib/filebench-prefix/src/filebench-stamp/filebench-update-*.log

**Comment by iomartin** [comment]
@mjlee34 you were right, that fixed the compilation issues on Ubuntu 18.04
